File: 787.cpp
class Solution {
public:
    // Ford-Bellman algo
    int findCheapestPrice(int n, std::vector<std::vector<int>>& flights, int src, int dst, int k) {
        std::vector<std::vector<int>> d(k + 2, std::vector<int>(n, INT_MAX));
        d[0][src] = 0;
        int u;
        int v;
        for (int i = 1; i <= k + 1; ++i) {
            for (int j = 0; j < flights.size(); ++j) {
                u = flights[j][0];
                v = flights[j][1];
                if (d[i - 1][u] != INT_MAX && d[i][v] > d[i - 1][u] + flights[j][2]) {
                    d[i][v] = d[i - 1][u] + flights[j][2];
                }
            }
        }
        return d[k + 1][dst] != INT_MAX ? d[k + 1][dst] : -1;
    }
};

File: 787.py
# This is the first solution that passes the time limit
# Previous ones are too slow, which makes them invalid

class Solution(object):
    def findCheapestPrice(self, n, flights, src, dst, k):
        prices = [float('inf')] * n
        prices[src] = 0
        queue = [src]
        next_queue = []
        tmp = list(prices)
        step = 0
        while queue and (step <= k):
            step += 1
            for flight in flights:
                if (flight[0] in queue) and (prices[flight[0]] + flight[2] < tmp[flight[1]]):
                    next_queue.append(flight[1])
                    tmp[flight[1]] = prices[flight[0]] + flight[2]
            prices = list(tmp)
            queue = list(next_queue)
            next_queue = []
        if type(prices[dst]) == int:
            return prices[dst]
        return -1
